.github/workflows/main.yml

```diff
@@ -2,40 +2,287 @@ name: CI
+          p12-file-base64: ${{ secrets.APPLE_CERTS_P12 }}
+          p12-password: ${{ secrets.APPLE_CERTS_PASS }}
+      - name: Verify identity used for signing
+        run: security find-identity -v
+      - name: Unpack any new darwin arm64 and amd64 binaries to ./tmp
+        run: |
+          # ./releases/{DIST_NAME}/{DIST_VERSION}/*_darwin-${arch}.tar.gz -> ./tmp/${DIST_NAME}_${DIST_VERSION}_${arch}-unsigned/
+          for NEW_DIR in ./releases/*/v*; do
+            (! test -d "$NEW_DIR") && continue
+            DIST_VERSION=$(basename "$NEW_DIR")
+            DIST_NAME=$(basename $(dirname "$NEW_DIR"))
+            for arch in "amd64" "arm64"; do
+              mkdir -p "./tmp/${DIST_NAME}_${DIST_VERSION}_${arch}-unsigned"
+              tar -zxvf "./releases/${DIST_NAME}/${DIST_VERSION}/${DIST_NAME}_${DIST_VERSION}_darwin-${arch}.tar.gz" -C "./tmp/${DIST_NAME}_${DIST_VERSION}_${arch}-unsigned/"
+            done
+          done
+          ls -Rhl ./tmp || echo "Nothing new in ./tmp"
+      - name: Sign and notarize the mac binaries
+        env:
+          AC_USERNAME: ${{ secrets.APPLE_AC_USERNAME }} # implicitly read from env by gon
+          AC_PASSWORD: ${{ secrets.APPLE_AC_PASSWORD }}
+        run: |
+          # Find and sign executables in ./tmp/${DIST_NAME}_${DIST_VERSION}_${arch}-unsigned/
+          for NEW_DIR in ./releases/*/v*; do
+            (! test -d "$NEW_DIR") && continue
+            DIST_VERSION=$(basename "$NEW_DIR")
+            DIST_NAME=$(basename $(dirname "$NEW_DIR"))
+            for arch in "amd64" "arm64"; do
+              EXECUTABLES=$(jq -nc '$ARGS.positional' --args $(find "./tmp/${DIST_NAME}_${DIST_VERSION}_${arch}-unsigned/" -perm +111 -type f -print))
+              echo "{
+                  \"source\" : $EXECUTABLES,
+                  \"bundle_id\" : \"io.ipfs.dist.${DIST_NAME}\",
+                  \"apple_id\": {
+                    \"password\":  \"@env:AC_PASSWORD\"
+                  },
+                  \"sign\" :{
+                    \"application_identity\" : \"Developer ID Application: Protocol Labs, Inc. (7Y229E2YRL)\"
+                  },
+                  \"zip\" :{
+                      \"output_path\" : \"./tmp/${DIST_NAME}_${DIST_VERSION}_${arch}-signed.zip\"
+                  }
+              }" | tee | jq > "./tmp/${DIST_NAME}_${DIST_VERSION}_${arch}-gon.json"
+              gon -log-level=info -log-json "./tmp/${DIST_NAME}_${DIST_VERSION}_${arch}-gon.json"
+            done
+          done
+      - name: Temporarily save ./tmp
+        uses: actions/upload-artifact@v2
+        with:
+          name: tmp
+          path: ./tmp/
+          retention-days: 1
+      - name: Update changed binaries in ./releases
+        run: |
+          for NEW_DIR in ./releases/*/v*; do
+            (! test -d "$NEW_DIR") && continue
+            DIST_VERSION=$(basename "$NEW_DIR")
+            DIST_NAME=$(basename $(dirname "$NEW_DIR"))
+            for arch in "amd64" "arm64"; do
```

Comment:
Is it normal to have big shell scripts inlined like this in GitHub workflow?  Or would it be better to factor it out?

(I don't know  conventions here).

Also, do we have a "3 finger rule" around bash (i.e., once bash code is more than 3 fingers of lines on the terminal, we use a different language that is more developer friendly)? (We don't need to introduce that here - just curious - feel free to disregard()

Reply:
I would support such a thing, but i don't believe there are any existing conventions for it.

Reply:
I feel shell scripts are the right tool for the job in CI workflows, but this could be my personal bias :)
Considered extracting those longer blocks to `.sh` files in  `./scripts/ci/` but it felt harder to read and reason about – left them inlined.

We can extract them after this + nightlies lands, but suggest doing that in a separate PR, to make reviews at hand easier by having everything in one place, removing need for jumps between files.

Reply:
Update: after reviewing #369 I believe I need to extract those bash blocks into reusable scripts to land everything anyway :) (want to run them in both `main.yml` for release / PR builds and `nightly.yml` for daily matrix build).
Stay tunned.

Reply:
Done. Extracted scripts are in `./scripts/ci` and will be shared with nightly  setup.